### PR TITLE
Update FilteredAdsQuery and user settings for opting-out of Billboards

### PR DESCRIPTION
**Pull request by jaw6**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Per https://github.com/forem/forem-internal-eng/issues/567, we want to update the user settings to limit what kinds of ads users can opt-out of seeing. 

221600284-11fb0349-0841-49e2-8055-2aa70d0b97e3

Previously, these settings applied to any/all ads, however, we want to limit these settings to apply only to `external`-type billboards. In the case of the first settings ("when browsing"), we'll hide `external`-type ads via JavaScript, meaning that users who have opted-out in this way will _sometimes_ see an `internal`- or `community`-type ad, but if the ad served up by `FilteredAdsQuery` is `external`, we'll hide it for them.

@pkfrank will provide updated copy before this can merge.

## Related Tickets & Documents

- Closes #https://github.com/forem/forem-internal-eng/issues/567

## QA Instructions, Screenshots, Recordings

* You'll want to have a local user who has articles published and at least one organization
* You'll want to have local billboards, at least one community, in-house, and external-type ad
* It can help to add `console.log` in `initializeDisplayAdVisibility` to confirm whether or not ads are being hidden
* Log-in as that user and disable one of these settings, then visit articles published by that user and articles not published by that user

### UI accessibility concerns?

Hidden ads on the client-side could conceivably still show up for folks who have JavaScript disabled, but that's likely a vanishingly small minority at this stage.

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/models/display_ad.rb [hunk 1]
```diff
@@ -37,14 +37,33 @@ class DisplayAd < ApplicationRecord
...
+    # evaluated" stuff a chance to get some engagement and start showing up more. If it doesn't get engagement, it
+    # stays in this area.
+
+    # Ads that get engagement have a higher "success rate", and among this category, we sample from the top 15 that
+    # meet that criteria. Within those 15 top "success rates" likely to be clicked, there is a weighting towards the
+    # top ranked outcome as well, and a steady decline over the next 15 — that's because it's not "Here are the top 15
+    # pick one randomly", it is actually "Let's cut off the query at a random limit between 1 and 15 and sample from
+    # that". So basically the "limit" logic will result in 15 sets, and then we sample randomly from there. The
+    # "first ranked" ad will show up in all 15 sets, where as 15 will only show in 1 of the 15.
+    if rand(8) == 1
```
Comment (jaw6) [inline app/models/display_ad.rb:62]: `cc` @Ridhwana @lboogie04 I'll explain a bit more below, but I wound up refactoring `filtered_ads_query_spec` quite a bit, and moving this randomization routine into the class here allowed those tests to be a _lot_ more explicit about what was being filtered, without worrying about false-(positive or negative) flaky tests that were passing one minute and failing the next due only to the randomness. 

We _might_ want to consider whether we add more tests to `display_ad_spec`, but testing this randomness is a bit brittle, at best.

File: app/queries/display_ads/filtered_ads_query.rb [hunk 2]
```diff
@@ -30,10 +32,9 @@ def call
                                 authenticated_ads(%w[all logged_out])
                               end
 
-      @filtered_display_ads = community_or_in_house_ads
+      @filtered_display_ads = type_matched_ads
```
Comment (jaw6) [inline app/queries/display_ads/filtered_ads_query.rb:35]: `cc` @Ridhwana @lboogie04 I renamed this method trying to reflect that it's also matching `external`-type ads and now dealing with `permit_adjacent_sponsors`. I'm not in love with this name, but it was the best I could think of that fit with the `_ads` pattern we have going here. I'm very open to suggestions.

File: spec/queries/display_ads/filtered_ads_query_spec.rb [hunk 3]
```diff
@@ -1,154 +1,125 @@
...
-                          placement_area: "post_comments",
-                          published: true,
-                          approved: true,
-                          cached_tag_list: "career")
-
-      article_tags = %w[productivity java]
-      expect(described_class.call(display_ads: DisplayAd.all, area: "post_comments", user_signed_in: false,
-                                organization_id: display_ad.organization.id, article_tags: article_tags)).to eq(display_ad)
+    # This original test name seems misleading, it does not do what it says
+    pending "will show no display ads if the available display ads have no tags set or do not contain matching tags" do
```
Comment (jaw6) [inline spec/queries/display_ads/filtered_ads_query_spec.rb:41]: `cc` @Ridhwana @lboogie04 It can be hard to follow in the diff here, but in the [original version of these tests](https://github.com/forem/forem/blob/cfad413ca131c1d0b7a2954d50779f04be335700/spec/queries/display_ads/filtered_ads_query_spec.rb#L47-L74) there are two tests with sort of contradictory set ups and the name of this one is — ¿I _think_? — a _bit_ misleading. 

As I understand the filter behavior, when there _are_ tags on the article, we would show ads that _match_ the tags or any ads that _have **no** tags_ (as ads without tags are intended to match _all_ articles, not intended to _only_ match tag-less articles — I _think_?) The original version of this test does pass with `nil` filtered ads, but I _think_ that's because the setup has (in effect) _only_ created one ad (with a mis-matched tag list).

I included this test (as `pending`) since I wanted to confirm that I understood any of this right — because it's also possible we've got a bug here somehow?

File: app/queries/display_ads/filtered_ads_query.rb [hunk 4]
```diff
@@ -59,33 +60,26 @@ def authenticated_ads(display_auth_audience)
...
+                    "(type_of = :community AND organization_id = :organization_id)"
+                  end
+
+      # If the article's author permits adjacent sponsors, we might show an external-type ad
+      # *if* the organization_id doesn't match the current article's organization id
+      external = if @permit_adjacent_sponsors && @organization_id
+                   "(type_of = :external AND organization_id != :organization_id)"
+                 elsif @permit_adjacent_sponsors
+                   "(type_of = :external)"
+                 end
```
Comment (jaw6) [inline app/queries/display_ads/filtered_ads_query.rb:78]: `cc` @Ridhwana @lboogie04 I'm not sure if this change is _really_ necessary, maybe it's fine, but I think database generally treat `NULL` in a way that makes it work strangely if it winds up in a `!= null` clause. ("NULL is a placeholder, not a value" is how I learned to think about it, but I'm no longer confident that any/all databases treat this the same way.)

**Comment by jaw6** [comment]
It's in the description (and I added the "blocked" label), but just to put it here as well: This should not merge until we have updated settings page copy from @pkfrank 

Screen Shot 2023-03-07 at 15 49 03

**Comment by pkfrank** [comment]
@jaw6 sorry to be an initial blocker, let's make this simple update for now:

> You have the option to disable billboards from external sponsors.  Our wonderful partners help sustain the platform, and we strive to ensure that their presence is constructive and valuable to the community, but you are welcome to make use of this setting if you wish.  More details in our [billboards FAQ](/billboards).

File: app/queries/display_ads/filtered_ads_query.rb (hunk 4 above)
Comment (Ridhwana) [reply app/queries/display_ads/filtered_ads_query.rb:78]: @jaw6 I was doing some testing this morning and this is definitely necessary, otherwise it was yielding the results whereby the ad was not showing up articles where the article `organization_id` was nil.

**Review by Ridhwana — APPROVED** [review]

File: app/queries/display_ads/filtered_ads_query.rb (hunk 2 above)
Comment (jaw6) [reply app/queries/display_ads/filtered_ads_query.rb]: `type_of_ads` is great, I'm going with that!
